### PR TITLE
Refactor MatchService to use message queues and account for edge cases

### LINKED ISSUE #24: [#24] [MatchService] [FR 2.2] [W6] Add checks for edge cases in MatchService
**Is your feature request related to a problem? Please describe.**
* User can have multiple find match requests at the same time
* User can match to himself
* User can cancel a find match request even though he did not create one

**Describe the solution you'd like**
Add checks to ensure that these scenarios don't happen

### PR DESCRIPTION
**Pull request by tau-bar**
Fixes #24, #17

Credit to @LZher385 for working on the timer on the frontend and creating the socket operations on `yuzher/match-ui` branch.

## Description
The edge cases have been accounted for. Verify using Postman by trying to match the same user with themselves/same socket with same socket, should not be able to find a match. On top of that, I have changed the structure of the MatchService to use RabbitMQ message queue.

## Type of change
- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
The tests are the same as before, the tests are on our shared Postman collection. 

`match:find`
Send this event with { "user": 1, "difficulty": "MEDIUM" } for example.

`match:cancel`
Send this event to cancel the current matchmaking.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by LZher385** [comment]
Match find and match cancel events tested and working. However I noticed that there is no event emitted when a user is deleted from the queue once the 30 seconds timer is out. Do you think there is a need for this? I guess the we can just start the 30s timer on the frontend separately as well so that it doesn't have to wait for such an event

**Comment by tyanhan** [comment]
image
Sending a match find request without difficulty causes the match service to be unable to start
image

File: matching-service/consumer/repository/matchRepository.js [hunk 1]
```diff
@@ -50,9 +50,25 @@ const deleteMatchByMatchInfo = async ({ user, difficulty, socketId }) => {
...
+	});
+};
+
+const deleteMatchBySocketId = async (socketId) => {
+	return await db.Match.destroy({
+		where: {
+			socketId: socketId,
+		}
+	});
+}
```
Comment (tyanhan) [inline matching-service/consumer/repository/matchRepository.js:65]: Missing `;` for the last function of most files

File: matching-service/package.json [hunk 2]
```diff
@@ -9,23 +9,25 @@
     "test": "echo \"Error: no test specified\" && exit 1",
     "lint": "npx eslint .",
     "lint-fix": "npx eslint . --fix",
-    "migrate": "node_modules/.bin/sequelize db:migrate"
+    "migrate": "./node_modules/.bin/sequelize db:migrate"
```
Comment (tyanhan) [inline matching-service/package.json:12]: The migrate function doesn't seem to be working on my side. It says '.' is not a recognized command. I was wondering if it is related to this. https://stackoverflow.com/questions/66185466/npm-script-fails-is-not-recognized-as-an-internal-or-external-command

**Review by tyanhan — CHANGES_REQUESTED** [review]
LGTM, just some minor changes

**Review by tyanhan — APPROVED** [review]
LGTM

**Pull request merged**